CPU Reservation

grandchild opened the pull request:
- vmware: Add CPU reservation support
- vmware: Update inventory with reservable CPU count

grandchild commented:
See also: https://github.com/sapcc/helm-charts/pull/5676

grandchild commented on nova/utils.py:
```diff
@@ -1238,26 +1241,32 @@ def is_numa_aligned_flavor(flavor):
+        else:
+            if value > 0:
+                return min(value, flavor_value)
+        if full_threshold >= 0 and flavor_value >= full_threshold:
```
you're right!

grandchild commented on nova/virt/vmwareapi/vm_util.py:
```diff
@@ -1646,23 +1646,28 @@ def aggregate_stats_from_cluster(host_stats):
-        CONF.vmware.memory_reservation_cluster_hosts_max_fail
+        CONF.vmware.resource_reservation_cluster_hosts_max_fail
     if max_fail_hvs and total_hypervisor_count:
+        vm_reservable_vcpu_ratio = \
```
hm you're right, that should be one value, at least here inside the function, if not also in the returned object.

grandchild replied:
i did it completely.

joker-at-work commented:
Why is fixing the tests its own commit? I expect that in porting to the next release, this will create additional work to merge it into the right commit(s) then, because we use the tests to figure out if the code should still work on the current branch. Therefore, I'd prefer the tests to work in both of these commits.

joker-at-work commented on nova/conf/vmware.py:
```diff
@@ -114,35 +114,38 @@
+resources would no longer be able to start.
 
 This setting applies to VMs with flavors which have a nonzero extra_spec
-"resources:CUSTOM_MEMORY_RESERVABLE_MB" set.
+"resources:CUSTOM_CPU_RESERVABLE" or "resources:CUSTOM_MEMORY_RESERVABLE_MB"
+set.
```
This sentence makes this setting sound like it's exclusively for VMs with these extra_specs. But that's not true, right? We also have a setting stating the memory over which every VM is related to this setting, too.

In general, it doesn't sound like this setting has anything to do with VMs but rather with the driver reporting resources. Not sure why we mention them here at all.

grandchild replied:
removed the paragraph in both configs.

joker-at-work commented on nova/conf/vmware.py:
```diff
@@ -114,35 +114,38 @@
+to start.
 
 This setting applies to VMs with flavors which have a nonzero extra_spec
-"resources:CUSTOM_MEMORY_RESERVABLE_MB" set.
+"resources:CUSTOM_CPU_RESERVABLE" or "resources:CUSTOM_MEMORY_RESERVABLE_MB"
+set.
```
Same here. The setting isn't really handled in VM contexts.

joker-at-work commented on nova/utils.py:
```diff
@@ -1238,26 +1241,32 @@ def is_numa_aligned_flavor(flavor):
-            return min(flavor.memory_mb, memory_mb)
+    def _get_reserved_from_flavor_and_conf(key, flavor_value, full_threshold):
+        # explicit definitions in the flavor take precedence over any heuristic
+        try:
+            value = int(flavor.extra_specs[key])
+        except (ValueError, KeyError):
+            return 0
+        else:
+            if value > 0:
+                return min(value, flavor_value)
+        if 0 <= full_threshold <= flavor_value:
+            return flavor_value
+        return 0
```
Would it make sense to pass in `flavor`, too, so that the function doesn't rely on being nested and able to access it from the "outside"?

Why do we `return 0` now instead of `pass`? This leads to us no longer handling `full_threshold` (or `CONF.full_reservation_memory_mb`) if the `key` does not exist. (Maybe write some tests for this method?)

grandchild replied:
passing in flavor would require us to pass in the attr for `flavor_value` as string, which i'm not a fan of (no type checking). the alternative would be to do `flavor.extra_specs[...SPEC_KEY]` on the outside, but then we lose the dry `try ... except KeyError` version.
so yeah, i get your point about the duplicate methods of parameter-passing, but i don't agree that i'd have to take care of being able to call the method from the outside. this method is there purely for dry-ness, not really for re-usability elsewhere.

i fixed the `pass`.

grandchild replied:
added some test cases

joker-at-work commented on nova/virt/vmwareapi/cluster_util.py:
```diff
@@ -429,3 +430,26 @@ def update_cluster_das_vm_override(session, cluster, vm_ref, operation='add',
+                                 "host")
+    if not hosts:
+        raise CpuReservationNoHostsError(cluster=cluster)
+    host_hz = [
+        session._call_method(vutil, "get_object_property", host,
+                             "hardware.cpuInfo.hz")
+        for host in hosts.ManagedObjectReference]
```
Please don't make an API call per host in the cluster. There's a method to get properties for a list of morefs.

grandchild replied:
obsoleted

joker-at-work commented on nova/virt/vmwareapi/cluster_util.py:
```diff
@@ -429,3 +430,26 @@ def update_cluster_das_vm_override(session, cluster, vm_ref, operation='add',
+                             "hardware.cpuInfo.hz")
+        for host in hosts.ManagedObjectReference]
+    host_hz.sort()
+    # NOTE(jakobk): Yes, the proper median for an even number of values is the
+    # average of the two middle values, but we don't need that precision.
+    # And yes, it's SI MHz, not MiHz.
```
Can we drop the `Yes` and `Any yes`? Sounds off to me.

grandchild replied:
obsoleted

joker-at-work commented on nova/virt/vmwareapi/cluster_util.py:
```diff
@@ -429,3 +430,26 @@ def update_cluster_das_vm_override(session, cluster, vm_ref, operation='add',
     reconfigure_cluster(session, cluster, config_spec)
+
+
+def get_cluster_median_cpu_mhz(session, cluster) -> int:
+    """Get the median CPU frequency for all hosts in a cluster.
```
I don't think `cluster_util` is used for this kind of functionality (which calculates something about hosts). Instead, I'd put it where we have the failover-host detection and such.

Btw. why do we have to fetch these things anew? Could we not take the `VCState` information for this? It should™ contain all the hosts and their cpu count. If we now also need the `hz`, maybe it would be best to add it there.

grandchild replied:
obsoleted

joker-at-work commented on nova/virt/vmwareapi/vmops.py:
```diff
@@ -422,10 +422,17 @@ def _get_extra_specs(self, flavor, image_meta=None):
+            if cpu_reserved > 0:
+                # NOTE(jakobk): CPU reservations are given in MHz, not in
+                # number-of-CPUs, so we have to translate here.
+                cluster_host_mhz = cluster_util.get_cluster_median_cpu_mhz(
+                                                self._session, self._cluster)
```
Do we have to query it on each spawn/resize or could we use some cached value (e.g. from `VCState`)?

joker-at-work commented on nova/virt/vmwareapi/cluster_util.py:
```diff
@@ -429,3 +430,26 @@ def update_cluster_das_vm_override(session, cluster, vm_ref, operation='add',
+    hosts = session._call_method(vutil, "get_object_property", cluster,
+                                 "host")
+    if not hosts:
+        raise CpuReservationNoHostsError(cluster=cluster)
```
Is `cluster` a full `moref`? If yes, I'd prefer the variable to show this either via `cluster_ref` or `cluster_ref_value` (if it's not a full MoRef). If it's a full MoRef, I'd also prefer `cluster=vutil.get_moref_value(cluster)` here, because a MoRef renders to string with multiple newlines or as JSON - depending on our `oslo.vwmare`.

grandchild replied:
obsoleted

grandchild commented:
sure, i can merge them in

joker-at-work reviewed (commented):
Missed 2 places for the config option rename:
```
$ git grep memory_reservation_cluster_hosts_max_fail
nova/conf/base.py:Compare the values of conf.vmware.memory_reservation_cluster_hosts_max_fail and

$ git grep memory_reservation_max_ratio_fallback
nova/conf/base.py:conf.vmware.memory_reservation_max_ratio_fallback to see how much of total
```

Why do we need the new resource? What happens with all the VCPUs from the VCPUs resource that are still available with 4-times overcommit but cannot be scheduled on the cluster because we reserved some VCPUs? I'm missing an explanation of the grand scheme of things in the commit-message. It only tells me details but not the why the resource and how it's supposed to be used and it's interaction with VCPUs and their allocation ratio.

grandchild commented on nova/conf/base.py:
```diff
@@ -128,15 +128,29 @@
         default=230 * 1024,      # 230 GiB
         help="""
-Instance memory usage identifying a VM as needing memory reservations
+Instance memory usage identifying a VM as needing memory reservations.
```
while i think that sentence fragments such as this, common at the start of docstrings, can absolutely have a period, i have removed it again. i didn't mean to touch this line at all.

grandchild commented on nova/utils.py:
```diff
@@ -1238,26 +1241,32 @@ def is_numa_aligned_flavor(flavor):
+    def _get_reserved_from_flavor_and_conf(key, flavor_value, full_threshold):
+        # explicit definitions in the flavor take precedence over any heuristic
+        try:
+            value = int(flavor.extra_specs[key])
```
okay i moved it out

grandchild commented:
- Changed the extra-spec from `CUSTOM_RESOURCE_...` to a new `reservation:cpu` because it's different than the custom reservable memory resource.
- Explained the reasoning more in the commit message.
- Removed the full-cpu-threshold config because we don't plan on using it atm. It's easy to add later if that changes.

joker-at-work commented:
Why do we have to set the number of cpus to reserve, if we expect this to only ever reserve all CPUs of the flavor? Couldn't we have a `reserve_vcpus`?

Did you check if there's already an upstream property for flavors to define cpu reservations?

grandchild commented on nova/virt/vmwareapi/vmops.py:
```diff
@@ -422,10 +422,17 @@ def _get_extra_specs(self, flavor, image_meta=None):
+            if cpu_reserved > 0:
+                # NOTE(jakobk): CPU reservations are given in MHz, not in
+                # number-of-CPUs, so we have to translate here.
+                cluster_host_mhz = \
+                    self._vc_state.get_host_stats().get('cpu_mhz', 0)
```
interesting. it did work but maybe that was a previous iteration of the code? checking...

grandchild replied:
fixed and tested in qa-de-2

grandchild commented:
> Why do we have to set the number of cpus to reserve, if we expect this to only ever reserve all CPUs of the flavor

a) it mirrors how we do it for memory
b) we might actually not reserve all CPUs for some reason or other in the future. unlikely, but possible.

> Did you check if there's already an upstream property for flavors to define cpu reservations?

No I did not. But there doesn't seem to be anything, except for, as you said, pinning and reservation for the hypervisor.

joker-at-work commented:
Could you please explain in the commit-message why we do not use `quota:cpu_reservation` as already supported by the vmwareapi driver upstream?

grandchild commented:
I added the blurb about `quota:cpu_reservation` to the commit message.
